### PR TITLE
simplify aggregators

### PR DESCRIPTION
**Pull request by cseed**
This implements my proposal to simplify aggregators here: http://dev.hail.is/t/proposal-for-aggregators/93/3.

Builds on: https://github.com/hail-is/hail/pull/3552

I still need to finish converting some tests from ExtractAggregatorSuite (they are currently commented out).

Old-style aggregators (filter, map, flatMap) are expanded at toIR conversion time to ApplyAggOp and SeqOp.  ApplyAggOp returns the result of the aggregation.  ApplyAggOp's first argument, which  must be of type TVoid, is the expression to run for each element being aggregated over.  SeqOp merges a computed value into the RegionValueAggregator.

I added a AggSignature that holds the AggOp, type being aggregated over and the RegionValueAggregator constructor argument types.  This is stored by the ApplyAggOp and the SeqOp.

@tpoterba I believe TAggregable is no longer used in the IR code and can go away when the AST gets ripped out.

@danking I added some IR testing logic to TestUtils.  Namely, `eval` evaluates an IR with environments, args and/or aggregations with a single call (and verifies that the interpret with and without optimization and compiler all agree).  There are also functions for asserting the result of aggregations, for example:

```
    val aggSig = AggSignature(Sum(), TFloat64(), FastSeq())
    assertEvalsTo(ApplyAggOp(
      SeqOp(ApplyBinaryPrimOp(Multiply(), Ref("a", TFloat64()), Ref("b", TFloat64())), I32(0), aggSig),
      FastSeq(), aggSig),
      (FastIndexedSeq(Row(1.0, 10.0), Row(10.0, 10.0), Row(null, 10.0)), TStruct("a" -> TFloat64(), "b" -> TFloat64())),
      110.0)
```

The line:

 >       (FastIndexedSeq(Row(1.0, 10.0), ...), TStruct("a" -> TFloat64(), "b" -> TFloat64())),

is the IndexedSeq of values to aggregate over, along with their signature.  The struct type is used to build the scope in which aggregators are evaluated.

(A little noisy because of the aggregator syntax.  It's noisier than I'd like it to be.)

This nicely paves the way for aggregators with multi-argument seqOps (like takeBy) that were previously handled by lambdas and we didn't have a clear plan for.

### CODE REVIEW

File: src/test/scala/is/hail/expr/ir/ExtractAggregatorsSuite.scala [hunk 1]
```diff
@@ -1,255 +1,108 @@
...
-      _ => 10.0)
-    assert(sum === 100.0)
+        SeqOp(Ref("a", TInt32()), I32(0), aggSig),
+        FastSeq(), aggSig),
+      (FastIndexedSeq(Row(10), Row(null), Row(5)), TStruct("a" -> TInt32())),
+      FastIndexedSeq(10, null, 5))
   }
 
+  /*
+
```
Comment (jigold) [inline src/test/scala/is/hail/expr/ir/ExtractAggregatorsSuite.scala]: I want to see a working `flatMap` test before I approve this. I assume you'll convert the rest of these.

File: src/main/scala/is/hail/expr/AST.scala [hunk 2]
```diff
@@ -853,8 +855,70 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
...
+          })
+        } yield rx
+      case "flatMap" =>
+        for {
+          bodyx <- body.toIR()
+          rx <- lhs.toAggIR(agg, { lhsx =>
+            val t = -lhsx.typ.asInstanceOf[TContainer].elementType
+            val v = ir.genUID()
+            ir.ArrayFor(lhsx, v,
+              ir.Let(name, ir.Ref(v, t), bodyx))
```
[the remark this answers is not in the document]
Comment (cseed) [reply src/main/scala/is/hail/expr/AST.scala]: Yep, good catch.

File: src/main/scala/is/hail/utils/MissingDoubleArrayBuilder.scala [hunk 3]
```diff
@@ -40,11 +40,14 @@ class MissingDoubleArrayBuilder {
   def write(rvb: RegionValueBuilder) {
     rvb.startArray(len)
     var i = 0
+    var j = 0
```
[the remark this answers is not in the document]
Comment (cseed) [reply src/main/scala/is/hail/utils/MissingDoubleArrayBuilder.scala]: I guess not.  Personally, I'd avoid higher order functions in data processing code like this, but it probably doesn't matter.  Let me know if you want me to clean it up.

File: src/main/scala/is/hail/expr/ir/Interpret.scala [hunk 4]
```diff
@@ -310,13 +315,22 @@ object Interpret {
...
+        xs.foreach(x => Interpret(x))
+      case x@SeqOp(a, i, aggSig) =>
+        assert(i == I32(0))
+        aggregator.get.seqOp(interpret(a))
+      case x@ApplyAggOp(a, constructorArgs, initOpArgs, aggSig) =>
+        val aggType = aggSig.inputType
+        assert(AggOp.getType(aggSig) == x.typ)
+        val aggregator = aggSig.op match {
+          case CallStats() =>
+            val nAlleles = interpret(initOpArgs.get(0))
```
Comment (jigold) [inline src/main/scala/is/hail/expr/ir/Interpret.scala]: assert (aggType == TCall())

File: src/main/scala/is/hail/utils/MissingDoubleArrayBuilder.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (jigold) [reply src/main/scala/is/hail/utils/MissingDoubleArrayBuilder.scala]: Leave it for now.

File: src/test/scala/is/hail/TestUtils.scala [hunk 5]
```diff
@@ -255,27 +256,40 @@ object TestUtils {
...
-  def eval(x: IR, args: IndexedSeq[(Any, Type)]): Any = {
-    val i = Interpret[Any](x, Env.empty[(Any, Type)], args, None)
+  def eval(x: IR, env: Env[(Any, Type)], args: IndexedSeq[(Any, Type)], agg: Option[(IndexedSeq[Row], TStruct)]): Any = {
+    val i = Interpret[Any](x, env, args, agg)
 
-    val i2 = Interpret[Any](x, Env.empty[(Any, Type)], args, None, optimize = false)
+    val i2 = Interpret[Any](x, env, args, agg, optimize = false)
     assert(i == i2)
 
+    // FIXME handle agg
```
Comment (jigold) [inline src/test/scala/is/hail/TestUtils.scala]: Is this fixme still needed?

File: src/main/scala/is/hail/expr/ir/AggOp.scala [hunk 6]
```diff
@@ -99,9 +103,22 @@ object AggOp {
...
+  val fromString: PartialFunction[String, AggOp] = {
+    case "fraction" => Fraction()
+    case "stats" => Statistics()
+    case "collect" => Collect()
+    case "sum" => Sum()
+    case "product" => Product()
+    case "max" => Max()
+    case "min" => Min()
+    case "count" => Count()
+    case "take" => Take()
```
[the remark this answers is not in the document]
Comment (cseed) [reply src/main/scala/is/hail/expr/ir/AggOp.scala]: Good catch. I had to make some additional changes to get callStats through toIR.  I added a Python test case and verified it is making it completely through the IR manually.

**Pull request merged**